onert-micro/onert-micro/include/core/OMHuffmanTranscoder.h

```diff
@@ -0,0 +1,177 @@
+  Node *p_right = nullptr;
+  T data;
+  unsigned int freq;
+};
+
+template <typename T> struct CompareNodes
+{
+  bool operator()(Node<T> *l, Node<T> *r) { return l->freq > r->freq; }
+};
+
+template <typename T> class HuffmanTranscoder
+{
+private:
+  Node<T> *root = nullptr;
+  std::unordered_map<T, std::string> huffmanCode;
+  std::vector<bool> encoded_bitset{};
+  std::size_t nodes_count = 0;
+
+private:
+  Node<T> *allocateNode(T data, unsigned int freq, Node<T> *p_left, Node<T> *p_right)
+  {
+    Node<T> *node = new Node<T>;
+    node->data = data;
+    node->freq = freq;
+    node->p_left = p_left;
+    node->p_right = p_right;
+    nodes_count++;
+    return node;
+  }
+  std::unordered_map<T, unsigned int> calculate_frequency_map(const std::vector<T> &input)
+  {
+    std::unordered_map<T, unsigned int> out_map;
+    for (auto &item : input)
+      out_map[item] = out_map.find(item) != out_map.end() ? out_map[item] + 1 : 1;
+    return out_map;
+  }
+
+  void buildHuffmanTable(Node<T> *node, const std::string str)
+  {
+    if (node == nullptr)
+      return;
+
+    if (!node->p_left && !node->p_right)
+    {
+      huffmanCode[node->data] = str;
+    }
+
+    buildHuffmanTable(node->p_left, str + "0");
+    buildHuffmanTable(node->p_right, str + "1");
+  }
+  void decode(Node<T> *node, int &index, std::string str)
+  {
+    if (node == nullptr)
+    {
+      return;
+    }
+
+    if (!node->p_left && !node->p_right)
+    {
+      std::cout << static_cast<int>(node->data);
+      return;
+    }
+
+    index++;
+
+    if (str[index] == '0')
+      decode(node->p_left, index, str);
+    else
+      decode(node->p_right, index, str);
+  }
+  void buildHuffmanTree(const std::vector<T> &input)
+  {
+    auto freq_map = calculate_frequency_map(input);
+
+    std::priority_queue<Node<T> *, std::vector<Node<T> *>, CompareNodes<T>> pq;
+
+    for (auto &item : freq_map)
+    {
+      pq.push(allocateNode(item.first, item.second, nullptr, nullptr));
+    }
+
+    while (pq.size() != 1)
+    {
+      Node<T> *left = pq.top();
+      pq.pop();
+      Node<T> *right = pq.top();
+      pq.pop();
+
+      unsigned int sum = left->freq + right->freq;
+      pq.push(allocateNode(0, sum, left, right));
+    }
+
+    root = pq.top();
+  }
+
+public:
+  void encode_decode_example()
+  {
+
+    std::vector<T> input;
+    for (int i = 0; i < 1024; ++i)
+      input.push_back(0);
+    for (int i = 0; i < 256; ++i)
+      input.push_back(i);
+
+    buildHuffmanTree(input);
+    buildHuffmanTable(root, "");
+
+    // std::cout << "Huffman Codes are :\n";
+    // for (auto pair : huffmanCode) {
+    //	std::cout << static_cast<int>(pair.first) << " " << pair.second << '\n';
+    // }
+
+    std::cout << "\nInput string bits:\n";
+    auto input_bits = input.size() * CHAR_BIT * sizeof(T);
+    std::cout << input_bits << "\n";
+
+    // TODO: replace string with bitset or bool vector
+    // print encoded string
+    std::string str = "";
```

Comment:
Maybe better rewrite it to bool or bitset in this draft? Since a string can take up a lot of extra memory, and then check memory after that?